File: src/csharp/examples/cudnn/Layer.cs
using Alea.CUDA;

namespace Tutorial.Cs.examples.cudnn
{

    public class Layer
    {
        public int Inputs { get; set; }
        public int Outputs { get; set; }
        public int KernelDim { get; set; }
        public float[] DataH { get; set; }
        public DeviceMemory<float> DataD { get; set; }
        public float[] BiasH { get; set; }
        public DeviceMemory<float> BiasD { get; set; }

        public void Dispose()
        {
            DataD.Dispose();
            BiasD.Dispose();
        }
        
        public static Layer CreateLayer(Worker worker, int inputs, int outputs, int kernelDim, string fnameWeights,string fnameBias)
        {
            var pathWeights = Data.GetPath(fnameWeights);
            var pathBias = Data.GetPath(fnameBias);
            var dataH = Data.ReadBinaryFile(pathWeights);
            var dataD = worker.Malloc(dataH);
            var biasH = Data.ReadBinaryFile(pathBias);
            var biasD = worker.Malloc(biasH);
            return new Layer
            {
                Inputs = inputs,
                Outputs = outputs,
                KernelDim = kernelDim,
                DataH = dataH,
                DataD = dataD,
                BiasH = biasH,
                BiasD = biasD
            };
        }

        public static Layer Conv1(Worker worker)
        {
            return CreateLayer(worker, 1, 20, 5, Data.Conv1Bin, Data.Conv1BiasBin);
        }

        public static Layer Conv2(Worker worker)
        {
            return CreateLayer(worker, 20, 50, 5, Data.Conv2Bin, Data.Conv2BiasBin);
        }

        public static Layer Ip1(Worker worker)
        {
            return CreateLayer(worker, 800, 500, 1, Data.Ip1Bin, Data.Ip1BiasBin);
        }

        public static Layer Ip2(Worker worker)
        {
            return CreateLayer(worker, 500, 10, 1, Data.Ip2Bin, Data.Ip2BiasBin);
        }
        
    }
    
}

File: src/fsharp/examples/cudnn/Layer.fs
(*** hide ***)
module Tutorial.Fs.examples.cudnn.Layer
open System
open System.IO
open Alea.CUDA

(*** define:CudnnMnistLayer ***)
type Layer = 
    {
        Inputs : int
        Outputs: int

        KernelDim : int

        DataH : float32[]
        DataD : DeviceMemory<float32>

        BiasH : float32[]
        BiasD : DeviceMemory<float32>
    }
        
    static member create (worker:Worker) inputs outputs kernelDim fnameWeights fnameBias =
        let weightsPath, biasPath = getPath fnameWeights, getPath fnameBias
        
        let dataH = readBinaryFile weightsPath (inputs*outputs*kernelDim*kernelDim)
        let dataD = worker.Malloc(dataH)
            
        let biasH = readBinaryFile biasPath outputs
        let biasD = worker.Malloc(biasH)
            
        { Inputs = inputs; Outputs = outputs; KernelDim = kernelDim; DataH = dataH; DataD = dataD; BiasH = biasH; BiasD = biasD }
    
    static member conv1 worker = Layer.create worker 1 20 5 Conv1Bin Conv1BiasBin
    static member conv2 worker = Layer.create worker 20 50 5 Conv2Bin Conv2BiasBin
    static member ip1 worker = Layer.create worker 800 500 1 Ip1Bin Ip1BiasBin
    static member ip2 worker = Layer.create worker 500 10 1 Ip2Bin Ip2BiasBin